jax/_src/lax/parallel.py

```diff
@@ -361,34 +367,80 @@ def pdot(x, y, axis_name, pos_contract=((), ()), pos_batch=((), ())):
 
 ### parallel primitives
 
+def _subst_all_names_in_param(
+    pname: str, params: core.ParamDict, subst: core.AxisSubst) -> core.ParamDict:
+  axis_name = params[pname]
+  if not isinstance(axis_name, (tuple, list)):
+    axis_name = (axis_name,)
+  result = dict(params)
+  result[pname] = sum(((name,) if isinstance(name, int) else subst(name)
+                       for name in axis_name),
+                      ())
+  return result
+
 def _allreduce_soft_pmap_rule(prim, reducer, vals, mapped, chunk_size,
-                              *, axis_name, axis_index_groups):
+                              *, axes, axis_index_groups):
   if axis_index_groups is not None:
     raise NotImplementedError("soft_pmap does not yet support axis_index_groups")
   reduced_vals = [reducer(x, [0]) if m else x for x, m in zip(vals, mapped)]
-  outs = prim.bind(*reduced_vals, axis_name=axis_name,
+  outs = prim.bind(*reduced_vals, axes=axes,
                    axis_index_groups=axis_index_groups)
   return outs, (False,) * len(vals)
 
+def _reduction_with_positional_batcher(prim, vals_in, dims_in, axis_index_groups,
+    transform_unmapped, transform_mapped):
+  if axis_index_groups is not None:
+    raise NotImplementedError("axis_index_groups not supported in vmap collectives. "
+                              "Please open a feature request!")
+  # TODO: Transpose all dims to 0, increment all axes
```

Comment:
I assume this is follow-up work.

Reply:
This is actually done. I just forgot to remove the TODO!